README.MD
# Toy Syntax Analyzer
By Anthony Nguyen and Jarod Nakamoto

# How to use
This program requires the use of four files:  `toy.l`, `toy.y`, a modification for the `y.tab.c` file (to print out shift and reduce actions) , and an input file.
It also requires lex and yacc. 

## Compilation instructions
```bash
lex toy.l
yacc toy.y -dv
perl -i -p0e 's/yybackup:.*goto yyreduce;/`cat changes`/sge' y.tab.c
gcc y.tab.c lex.yy.c -lfl
./a.out < inputfile.txt
```

* Compile `toy.l` using lex. This should create `lex.yy.c`
* Compile `toy.y` using yacc. This should create `y.tab.c`, `y.tab.h`, and `y.output`
* Use perl to modify the `y.tab.c` file with the `changes` file to allow it to print shift and reduce actions
* Then use command `gcc toy.c lex.yy.c -o toy` to output to `toy`
* Run `a.out` while using an input file

Alternatively a file named `compile.sh` has been provided that will automatically create the `a.out` file.

```bash
./compile.sh
./a.out < inputfile.txt
```


changes
//Changes made to y.tab.c file
int test = 0;
static const char *const names[] =
{
  "$end", "error", "$undefined", "boolean", "break", "class",
  "double", "else", "extends", "for", "if", "implements", "int",
  "interface", "new", "newarray", "null", "println", "readln",
  "return", "string", "void", "while", "plus", "minus",
  "multiplication", "division", "mod", "less", "lessequal",
  "greater", "greaterequal", "equal", "notequal", "and", "or",
  "not", "assignop", "semicolon", "comma", "period", "leftparen",
  "rightparen", "leftbracket", "rightbracket", "leftbrace",
  "rightbrace", "intconstant", "doubleconstant", "stringconstant",
  "booleanconstant", "id"
};
yybackup:

/* Do appropriate processing given the current state.  Read a
   lookahead token if we need one and don't already have one.  */

/* First try to decide what to do without reference to lookahead token.  */
yyn = yypact[yystate];
if (yypact_value_is_default (yyn))
  goto yydefault;

/* Not known => get a lookahead token if don't already have one.  */

/* YYCHAR is either YYEMPTY or YYEOF or a valid lookahead symbol.  */
if (yychar == YYEMPTY)
  {
    YYDPRINTF ((stderr, "Reading a token: "));
    yychar = yylex ();
  test = 1; //EDIT
  }

if (yychar <= YYEOF)
  {
    yychar = yytoken = YYEOF;
    YYDPRINTF ((stderr, "Now at end of input.\n"));
  }
else
  {
    yytoken = YYTRANSLATE (yychar);
  if (test)
        printf("\n%s ",names[yytoken]); //EDIT
    test = 0;
    YY_SYMBOL_PRINT ("Next token is", yytoken, &yylval, &yylloc);
  }

  /* If the proper action on seeing token YYTOKEN is to reduce or to
     detect an error, take that action.  */
  yyn += yytoken;
  if (yyn < 0 || YYLAST < yyn || yycheck[yyn] != yytoken)
    goto yydefault;
  yyn = yytable[yyn];
  if (yyn <= 0)
    {
      if (yytable_value_is_error (yyn))
        goto yyerrlab;
      yyn = -yyn;
      goto yyreduce;
    }

  /* Count tokens shifted since error; after three, turn off error
     status.  */
  if (yyerrstatus)
    yyerrstatus--;

  /* Shift the lookahead token.  */
  YY_SYMBOL_PRINT ("Shifting", yytoken, &yylval, &yylloc);
  if (yytoken == 0) //EDIT
	  printf("\n[accept]\n");
  else
  	printf("[shift]");
  /* Discard the shifted token.  */
  yychar = YYEMPTY;

  yystate = yyn;
  YY_IGNORE_MAYBE_UNINITIALIZED_BEGIN
  *++yyvsp = yylval;
  YY_IGNORE_MAYBE_UNINITIALIZED_END

  goto yynewstate;


/*-----------------------------------------------------------.
| yydefault -- do the default action for the current state.  |
`-----------------------------------------------------------*/
yydefault:
  yyn = yydefact[yystate];
  if (yyn == 0){
    printf("%s\n","\n[reject]" );
    goto yyerrlab;
  }
  printf("[reduce %d]",yyn-1); //EDIT
  goto yyreduce;




toy.h
#define _boolean 1
#define _break 2
#define _class 3
#define _double 4
#define _else 5
#define _extends 6
#define _for 7
#define _if 8
#define _implements 9
#define _int 10
#define _interface 11
#define _new 12
#define _newarray 13
#define _null 14
#define _println 15
#define _readln 16
#define _return 17
#define _string 18
#define _void 19
#define _while 20
#define _plus 21
#define _minus 22
#define _multiplication 23
#define _division 24
#define _mod 25
#define _less 26
#define _lessequal 27
#define _greater 28
#define _greaterequal 29
#define _equal 30
#define _notequal 31
#define _and 32
#define _or 33
#define _not 34
#define _assignop 35
#define _semicolon 36
#define _comma 37
#define _period 38
#define _leftparen 39
#define _rightparen 40
#define _leftbracket 41
#define _rightbracket 42
#define _leftbrace 43
#define _rightbrace 44
#define _intconstant 45
#define _doubleconstant 46
#define _stringconstant 47
#define _booleanconstant 48
#define _id 49


y.tab.h
/* A Bison parser, made by GNU Bison 3.0.5.  */

/* Bison interface for Yacc-like parsers in C

   Copyright (C) 1984, 1989-1990, 2000-2015, 2018 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* As a special exception, you may create a larger work that contains
   part or all of the Bison parser skeleton and distribute that work
   under terms of your choice, so long as that work isn't itself a
   parser generator using the skeleton or a modified version thereof
   as a parser skeleton.  Alternatively, if you modify or redistribute
   the parser skeleton itself, you may (at your option) remove this
   special exception, which will cause the skeleton and the resulting
   Bison output files to be licensed under the GNU General Public
   License without this special exception.

   This special exception was added by the Free Software Foundation in
   version 2.2 of Bison.  */

#ifndef YY_YY_Y_TAB_H_INCLUDED
# define YY_YY_Y_TAB_H_INCLUDED
/* Debug traces.  */
#ifndef YYDEBUG
# define YYDEBUG 0
#endif
#if YYDEBUG
extern int yydebug;
#endif

/* Token type.  */
#ifndef YYTOKENTYPE
# define YYTOKENTYPE
  enum yytokentype
  {
    _boolean = 258,
    _break = 259,
    _class = 260,
    _double = 261,
    _else = 262,
    _extends = 263,
    _for = 264,
    _if = 265,
    _implements = 266,
    _int = 267,
    _interface = 268,
    _new = 269,
    _newarray = 270,
    _null = 271,
    _println = 272,
    _readln = 273,
    _return = 274,
    _string = 275,
    _void = 276,
    _while = 277,
    _plus = 278,
    _minus = 279,
    _multiplication = 280,
    _division = 281,
    _mod = 282,
    _less = 283,
    _lessequal = 284,
    _greater = 285,
    _greaterequal = 286,
    _equal = 287,
    _notequal = 288,
    _and = 289,
    _or = 290,
    _not = 291,
    _assignop = 292,
    _semicolon = 293,
    _comma = 294,
    _period = 295,
    _leftparen = 296,
    _rightparen = 297,
    _leftbracket = 298,
    _rightbracket = 299,
    _leftbrace = 300,
    _rightbrace = 301,
    _intconstant = 302,
    _doubleconstant = 303,
    _stringconstant = 304,
    _booleanconstant = 305,
    _id = 306,
    IFX = 307,
    _uminus = 308
  };
#endif
/* Tokens.  */
#define _boolean 258
#define _break 259
#define _class 260
#define _double 261
#define _else 262
#define _extends 263
#define _for 264
#define _if 265
#define _implements 266
#define _int 267
#define _interface 268
#define _new 269
#define _newarray 270
#define _null 271
#define _println 272
#define _readln 273
#define _return 274
#define _string 275
#define _void 276
#define _while 277
#define _plus 278
#define _minus 279
#define _multiplication 280
#define _division 281
#define _mod 282
#define _less 283
#define _lessequal 284
#define _greater 285
#define _greaterequal 286
#define _equal 287
#define _notequal 288
#define _and 289
#define _or 290
#define _not 291
#define _assignop 292
#define _semicolon 293
#define _comma 294
#define _period 295
#define _leftparen 296
#define _rightparen 297
#define _leftbracket 298
#define _rightbracket 299
#define _leftbrace 300
#define _rightbrace 301
#define _intconstant 302
#define _doubleconstant 303
#define _stringconstant 304
#define _booleanconstant 305
#define _id 306
#define IFX 307
#define _uminus 308

/* Value type.  */
#if ! defined YYSTYPE && ! defined YYSTYPE_IS_DECLARED
typedef int YYSTYPE;
# define YYSTYPE_IS_TRIVIAL 1
# define YYSTYPE_IS_DECLARED 1
#endif


extern YYSTYPE yylval;

int yyparse (void);

#endif /* !YY_YY_Y_TAB_H_INCLUDED  */


toy.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "y.tab.h"

extern int yylex();
extern int yylineno;
extern char* yytext;

char *tokens[] = {NULL, "boolean", "break", "class", "double", "else", "extends", "for", "if", "implements", "int", "interface", "new", "newarray", "null", "println", "readln", "return", "string", "void", "while", "plus", "minus", "multiplication", "division", "mod", "less", "lessequal", "greater", "greaterequal", "equal", "notequal", "and", "or", "not", "assignop", "semicolon", "comma", "period", "leftparen", "rightparen", "leftbracket", "rightbracket", "leftbrace", "rightbrace", "intconstant", "doubleconstant", "stringconstant", "booleanconstant", "id" };

// adds string to trie structure
int add_to_trie (char*, int*, char*, int*, int*);

// gets next char of the string
char getNextSymbol(char*,int*);

// add new string to trie structure
int create(char,int*, char*, char*, int*, int*);

// return switch position of character
int symPos(char);

// function to print out trie table
int printTrie(int* switchArr, char* symbolArr, int* nextArr, int* cur);

int main(void)
{
	int CHARS = 52;
	int MAXTRANSITION = 4096;
	int cur;
	cur = 0;
	int switchArr[CHARS]; 
	char symbolArr[MAXTRANSITION];
	int nextArr[MAXTRANSITION];
	int ntoken;
	int offset = _boolean-1;

	for(int i = 0; i < CHARS ; i++ ){
		switchArr[i] = -1;
	}

	ntoken = yylex();
	while(ntoken) {
		printf("%s ",tokens[ntoken-offset]);
		char input[strlen(yytext)]; 
		strcpy(input,yytext);
		if (ntoken == _id || ntoken <= _while || ntoken == _booleanconstant){
			add_to_trie(input, switchArr, symbolArr, nextArr, &cur);
		}
		ntoken = yylex();
	}
	printTrie(switchArr,symbolArr,nextArr,&cur);
	return 0;
}

int symPos(char x){
	if ( x >= 65 && x <= 90 )
	       	return x-65;
	else 
		return x-97+26;
}

char getNextSymbol(char* input, int* pos){
	int old = *pos;
	*pos = *pos + 1;
	return input[old];
}

int create(char valueOfSymbol, int* switchArr, char* symbolArr, char* input, int* pos, int* cur ){
	int exit = 0;
	int size = strlen(input);
	char nextSym;
	while( !exit ){
		if(size > 1){
		nextSym = getNextSymbol(input,pos);
		symbolArr[*cur] = nextSym;
		*cur = *cur + 1;
		}
		if (*pos >= size){
			symbolArr[*cur] = '*';
			*cur = *cur + 1;
			exit = 1;
		}
	}
	return 0;
}

int add_to_trie(char* input, int* switchArr, char* symbolArr, int* nextArr, int* cur){
	char endMarker = '*';
	int pos;
	pos = 0;
	
	char valueOfSymbol = getNextSymbol(input,&pos);
	valueOfSymbol = input[0];
	int ptr = switchArr[symPos(valueOfSymbol)];
	if (ptr == -1){
		switchArr[symPos(valueOfSymbol)] = *cur;
		create(valueOfSymbol, switchArr, symbolArr, input, &pos, cur);
	}
	else
	{
		if ( strlen(input) == 1){
			while (nextArr[ptr] > 0){
				ptr = nextArr[ptr];
			}
			if (symbolArr[ptr] == endMarker)
					return 0;
			if(symbolArr[*cur] != endMarker){
				nextArr[ptr] = *cur;
				symbolArr[*cur] = endMarker;
				*cur = *cur + 1;
			}
			return 0;
		}
		valueOfSymbol = getNextSymbol(input,&pos);
		int exit = 0;
		while ( !exit )
		{
			if ( strlen(input) == 2 && valueOfSymbol == symbolArr[ptr])
				return 0;
			if (symbolArr[ptr] == valueOfSymbol)
				if(valueOfSymbol != endMarker)
				{
					ptr = ptr + 1;
					valueOfSymbol = getNextSymbol(input,&pos);
					if( pos == strlen(input) )
					{
						if ( symbolArr[ptr+1] == endMarker){
							exit = 1;
							break;
						}
						int loop = 1;
						while ( loop ){
						if (nextArr[ptr] == 0){
							if ( symbolArr[ptr] == valueOfSymbol ) 
								nextArr[ptr+1] = *cur;
							else
								nextArr[ptr] = *cur;
							loop = 0;
						}
						else
							ptr = nextArr[ptr];
						}
						if ( symbolArr[ptr] != valueOfSymbol ) {
						symbolArr[*cur] = valueOfSymbol;
						*cur = *cur + 1;
						}
						symbolArr[*cur] = endMarker;
						*cur = *cur + 1;
						exit = 1;
					}
				}
				else
					exit = 1;
			else if (nextArr[ptr] > 0)
				ptr = nextArr[ptr];
			else
			{
				nextArr[ptr] = *cur;
				pos--;
				create(valueOfSymbol, switchArr, symbolArr, input, &pos, cur);
				exit = 1;
			}
		} //while
	} //if

	return 0;
}

int printTrie(int* switchArr, char* symbolArr, int* nextArr, int* cur){
	int space = 4;
	printf("\n\t");
	for( int i = 'A'; i <= 'T'; i++ ){
		printf("%*c",space,i);
	}
	printf("\nswitch:\t");
	for( int i = 0; i < 20; i++ ){
		printf("%*d",space,switchArr[i]);
	}
	printf("\n\n\t");
	for( int i = 'U'; i <= 'Z'; i++ ){
		printf("%*c",space,i);
	}
	for( int i = 'a'; i <= 'n'; i++ ){
		printf("%*c",space,i);
	}
	printf("\nswitch:\t");
	for( int i = 20; i < 40; i++ ){
		printf("%*d",space,switchArr[i]);
	}
	printf("\n\n\t");
	for( int i = 'o'; i <= 'z'; i++ ){
		printf("%*c",space,i);
	}
	printf("\nswitch:\t");
	for( int i = 40; i < 52; i++ ){
		printf("%*d",space,switchArr[i]);
	}
	printf("\n\n");


	for ( int j = 0; j < *cur/20+1; j++){
	if (*cur%20 == 0 && j == *cur/20)
		continue;		
	printf("\n\t");
	for ( int i = 0; i < 20; i++ ){
		if (i+j*20 < *cur)
		printf("%*d",space,i+j*20);
	}
	printf("\nsymbol:\t");
	for ( int i = 0; i < 20; i++ ){
		//if(symbolArr[i+j*20] == NULL)
		//	symbolArr[i+j*20] = '.';
		if (i+j*20 < *cur)
		printf("%*c",space,symbolArr[i+j*20]);
	}
	printf("\nnext:\t");
	for ( int i = 0; i < 20; i++ ){
		if (i+j*20 < *cur){
		if ( nextArr[i+j*20] > 0 )
			printf("%*d",space,nextArr[i+j*20]);
		else
			printf("%*c",space,' ');
		}
	}
	printf("\n");
	}

	//printf("SIZE:%d\n",*cur);
	return 0;
}

